### ISSUE TITLE
Regression in MoltenVK 1.1.9 affecting DXVK games

### ISSUE THREAD

**Issue by FleetAdmiralButter** [KhronosGroup/MoltenVK#1592]
It appears that release 1.1.9 has some significant regressions when running Final Fantasy XIV under Wine compared to 1.1.8:

![Screen Shot 2022-05-11 at 9 53 05 pm](https://user-images.githubusercontent.com/37036084/167843943-3bd773a1-ccc9-4c9d-b284-e4de1bb452f2.png)

The game is pretty much rendered unplayable with this bug - all the colours appear incorrectly and it will eventually lock up when loading into a character.

**Comment by seathasky** [comment]
Also experienced this on Final Fantasy XIV with 1.1.9.

**Comment by FleetAdmiralButter** [comment]
Reverting https://github.com/KhronosGroup/MoltenVK/commit/fada8e08c7a9868003992a087474e23b8c096142 seems to resolve the issue in both FFXIV as well as other games running under Wine/DXVK.

**Comment by rcaridade145** [comment]
Isn't this fixed by the corrections made to fix  https://github.com/KhronosGroup/MoltenVK/issues/1569 ?

**Comment by billhollings** [comment]
> Isn't this fixed by the corrections made to fix #1569 ?

Unfortunately, the 1.1.9 release was cut before that #1573 regression fix went in.

Can someone please test this issue against the latest MoltenVK version, or any version later than 6b3ef7b51a1793cdc268b5d5d9d96c5f8606e822, please?

**Comment by ZoeyCluff** [comment]
Completely broken in new and exciting ways!
[CleanShot 2022-05-11 at 16 13 19](https://user-images.githubusercontent.com/2716914/167948827-cd1dd608-74fa-4297-9f45-f778bd26cf14.png)
![image](https://user-images.githubusercontent.com/2716914/167948995-da854332-698b-4554-8813-b5e123f63020.png)

**Comment by Gcenx** [comment]
The following env options are being set
```
MVK_ALLOW_METAL_FENCES=1
MVK_CONFIG_RESUME_LOST_DEVICE=1
MVK_CONFIG_FULL_IMAGE_VIEW_SWIZZLE=1
```

Along with this copy of [DXVK-macOS](https://github.com/Gcenx/DXVK-macOS/releases/download/v1.10.1/dxvk-macOS-async-1.10.1.tar.gz) (and yes this contains the hacks from DXVK-CX21)

**Comment by ZoeyCluff** [comment]
videos:
https://www.youtube.com/watch?v=s8lleqBHLOk
https://www.youtube.com/watch?v=_gudqCgTG0I

**Comment by Gcenx** [comment]
@FleetAdmiralButter could you rename this since were aware (Wineskin/PortingKit) this affects all titles using DXVK

100% affects;
- GTA5
- Witcher 3
Didn’t continue to testing other titles

**Comment by reckenrode** [comment]
I’m not sure how helpful this is, but I thought I’d share my experience here that I mentioned in the other issue.

I’m _not_ experiencing issues with MoltenVK 1.1.9 in FFXIV. I’m playing on Wine 7.8 from nixpkgs along with DXVK (with [async][1], [compat][2], and [threading][3] patches) and MoltenVK (with [compat][4] patch) also from nixpkgs. As I mentioned in the other thread, the only unusual thing that’s being done in the DXVK build is nixpkgs uses [mcfgthread][5] for its C++ threading implementation.

[1]: https://github.com/Sporif/dxvk-async
[2]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/misc/dxvk/darwin-dxvk-compat.patch
[3]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/misc/dxvk/darwin-thread-primitives.patch
[4]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/os-specific/darwin/moltenvk/dxvk-moltenvk-compat.patch
[5]: https://github.com/lhmouse/mcfgthread

**Comment by Gcenx** [comment]
I’ve seen the white screen problem with Witcher3/GTA5 along with other titles when using MoltenVK-1.1.9 this using Xcode to build from source of using GitHub actions builds. @FleetAdmiralButter also experienced this issue with there own builds.

@reckenrode I’m wondering if there’s something with how MoltenVK is building built via nix that’s avoiding this bug.

**Comment by ZoeyCluff** [comment]
did you need to manually patch dxvk to run it? I've honestly never used nix before.

**Comment by reckenrode** [comment]
> @reckenrode I’m wondering if there’s something with how MoltenVK is building built via nix that’s avoiding this bug.

[This][1] is the derivation. It’s impure, so it’s using the system Xcode. The build process is a bit hacked up (to eventually support xcbuild once the x86_64-darwin SDK in nixpkgs is new enough), but it’s not applying any patches except the above for use with DXVK. The only dependency it doesn’t override (with the git commits in `ExternalRevisions`) is cereal. I use the version of cereal in nixpkgs because it has a patch for [CVE-2020-11105][cve] (which is fixed upstream in 1.3.1). 

> @ZoeyCluff did you need to manually patch dxvk to run it? I've honestly never used nix before.

The derivation takes care of that, but yes. It applies the patches to DXVK then builds it from source. You should be able to apply them to the release and build manually. If you do want to try with nix, see below.

Note: You’ll need Xcode installed. Also untested (other than that the builds succeed) since I have everything set up in a [derivation][2] in my nix configs.

After [installing nix][3], run the following. The commands are verbose because it avoids needing to configure anything once the installation is done.

```
$ nix --extra-platforms x86_64-darwin --extra-experimental-features "flakes nix-command" build --system x86_64-darwin github:NixOS/nixpkgs/nixpkgs-unstable#dxvk.bin 
$ nix --extra-platforms x86_64-darwin --extra-experimental-features "flakes nix-command" build --system x86_64-darwin --impure --expr 'let pkgs = (builtins.getFlake "github:NixOS/nixpkgs/nixpkgs-unstable").legacyPackages.x86_64-darwin; in pkgs.dxvk.patchMoltenVK pkgs.darwin.moltenvk'
$ nix --extra-platforms x86_64-darwin --extra-experimental-features "flakes nix-command" build --system x86_64-darwin github:NixOS/nixpkgs/nixpkgs-unstable#pkgsCross.mingwW64.windows.mcfgthreads --out-link mcfgthread
```

The DXVK DLLs will be in `./result/x64`, MoltenVK will be in `./result/lib/`, and `mcfgthread-12.dll` will be in `./mcfgthread/bin`. You’ll need to copy `mcfgthread-12.dll` along with the DXVK DLLs because they’re linked against it.

If you want to use my derivation, which takes care of everything, just do the following. Note that it will have to compile Wine, which takes a while. It also assumes FFXIV is installed in XDG directories*, so you’ll need to link FFXIV from another Wine prefix to `$HOME/.local/share/ffxiv/drive_c/Program Files/FFXIV`. DXVK logs are in `$HOME/.local/state/ffxiv`. FFXIV settings are in `$HOME/.config`. DXVK state cache is in `$HOME/.cache/ffxiv`.

```
nix --extra-platforms x86_64-darwin --extra-experimental-features "flakes nix-command" build --system x86_64-darwin github:reckenrode/nixos-configs#ffxiv --out-link ffxiv-nix
```

[1]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/os-specific/darwin/moltenvk/default.nix
[2]: https://github.com/reckenrode/nixos-configs/blob/main/pkgs/ffxiv/default.nix
[3]: https://nixos.org/download.html
[cve]: https://nvd.nist.gov/vuln/detail/CVE-2020-11105

**Comment by ZoeyCluff** [comment]
@reckenrode submitted a PR - your derivation is pulling a removed ffxiv download (1.0.5 has been silently removed and released with an unannounced 1.0.7)
https://github.com/reckenrode/nixos-configs/pull/2

**Comment by reckenrode** [comment]
> @reckenrode submitted a PR - your derivation is pulling a removed ffxiv download (1.0.5 has been silently removed and released with an unannounced 1.0.7)
> [reckenrode/nixos-configs#2](https://github.com/reckenrode/nixos-configs/pull/2)

Thanks. The hash needs updated too. I left feedback on the PR with the correct hash.

**Comment by reckenrode** [comment]
After some more testing, I can’t seem to break it on my system. Also tried CrossOver with MoltenVK built without nix and DXVK-macOS. Wish I could contribute something more useful than that. 😕

**Comment by FleetAdmiralButter** [comment]
> > Isn't this fixed by the corrections made to fix #1569 ?
> 
> Unfortunately, the 1.1.9 release was cut before that #1573 regression fix went in.
> 
> Can someone please test this issue against the latest MoltenVK version, or any version later than [6b3ef7b](https://github.com/KhronosGroup/MoltenVK/commit/6b3ef7b51a1793cdc268b5d5d9d96c5f8606e822), please?

@billhollings, we just tried out a build of MoltenVK using the tip of `master` (27aaaf50c1ad1922561a8785c3a99e19903951fb), but the regression still persists on both M1 and AMD GPUs.

**Comment by billhollings** [comment]
PR #1595 should fix this. Please retest using latest MoltenVK and close this if the issue is fixed.

**Comment by Gcenx** [comment]
@billhollings grabbed the archive from actions and witcher3 is still flickering like crazy

**Comment by FleetAdmiralButter** [comment]
@billhollings thanks for this! #1592 has eliminated the corruption on the title screen but unfortunately it's still present once you load into the game:

[Screen Recording 2022-05-20 at 11.45.59 pm.mov](https://youtu.be/uZzj2fN3WGE)

**Comment by billhollings** [comment]
> once you load into the game

Thanks for reporting. Looking at this now.

I'm not really familiar with FFXIV as a user. Is there a way to bypass the world login queues and long intro cutscenes to get to areas that are experiencing the problem? Otherwise, it make debugging painfully slow.

I'm testing with XIV on Mac.

**Comment by billhollings** [comment]
Also...if someone can help me figure out why `~/Library/Application Support/XIV on Mac/wine.log` doesn't include logs made by MoltenVK during the game, I'd appreciate it. The log sometimes contains the initial `VkInstance` report text, but usually doesn't. And it never includes any MoltenVK logging after that.

**Comment by FleetAdmiralButter** [comment]
> I'm not really familiar with FFXIV as a user. Is there a way to bypass the world login queues and long intro cutscenes to get to areas that are experiencing the problem? Otherwise, it make debugging painfully slow.

Unfortunately not, FFXIV has gained a lot of popularity recently and there is always a queue to load into the game at peak times.

> Also...if someone can help me figure out why ~/Library/Application Support/XIV on Mac/wine.log doesn't include logs made by MoltenVK during the game, I'd appreciate it. The log sometimes contains the initial VkInstance report text, but usually doesn't. And it never includes any MoltenVK logging after that.

You can change the Wine logging level in Preferences > Advanced > Wine Debug

**Comment by Gcenx** [comment]
To bypass the videos you need to edit FFXIV.cfg

tbats in your users documents/my games/final fantasy (blah)/ffxiv.cfg

now add the entry 

cutsceneMovieOpening 1


For expanded Vulkan logging try adding +vulkan to the winedebug field,

**Comment by billhollings** [comment]
> You can change the Wine logging level in Preferences > Advanced > Wine Debug

Hmmm... `Wine Debug` preference is already set to `-all`. But nothing from MoltenVK is showing up there.

**Comment by ZoeyCluff** [comment]
@billhollings might be easier to come in the discord if you use it. https://discord.gg/xivmac

**Comment by ZoeyCluff** [comment]
> > You can change the Wine logging level in Preferences > Advanced > Wine Debug
> 
> Hmmm... `Wine Debug` preference is al

ready set to `-all`. But nothing from MoltenVK is showing up there.

you need +all for that. 
![CleanShot 2022-05-20 at 10 12 43](https://user-images.githubusercontent.com/2716914/169558517-8829feee-ebe4-4992-9d6a-e7b80b388ee8.png)

**Comment by Gcenx** [comment]
> > You can change the Wine logging level in Preferences > Advanced > Wine Debug
> 
> Hmmm... `Wine Debug` preference is already set to `-all`. But nothing from MoltenVK is showing up there (other than occasionally the initial `VkInstance` entry).

-all removed all debugging remove that and possibly add +vulkan

**Comment by Gcenx** [comment]
If only there wasn’t the lighting issue in FFXIV I’d have recommended some other titles to use.

**Comment by billhollings** [comment]
> @billhollings thanks for this! #1592 has eliminated the corruption on the title screen but unfortunately it's still present once you load into the game:
> 
> [Screen Recording 2022-05-20 at 11.45.59 pm.mov](https://youtu.be/uZzj2fN3WGE)

I've updated this PR to properly handle the app possibly binding extra vertex attribute buffers that are not used by the current pipeline, but overlap the buffer binding range used by the implicit buffers needed by the current pipeline. From my initial testing, this seems to fix the game play portion as well.

Please retest with this updated PR. Thanks.

**Comment by ZoeyCluff** [comment]
building and testing now.

**Comment by MysticalOS** [comment]
hoping that also fixes witcher 3 white flicker. using 1.1.8 for now til that regression also worked out. Probably all tied together.

**Comment by Gcenx** [comment]
Not sure about fixing FFXIV but this didn't fix Witcher3

**Comment by ZoeyCluff** [comment]
Nope, worse this time.
![CleanShot 2022-05-22 at 15 19 52](https://user-images.githubusercontent.com/2716914/169714235-40697b63-5ddb-4c74-a28c-a851f9851555.png)
![CleanShot 2022-05-22 at 15 20 13](https://user-images.githubusercontent.com/2716914/169714246-a94bcc22-97bc-4a8f-a0a5-9c3225e10ab6.png)


https://nextcloud.zoeyrae.dev/s/zGFKGkrS8ExxaCy  Is the wine.log from the attempt.

**Comment by ZoeyCluff** [comment]
> Nope, worse this time. ![CleanShot 2022-05-22 at 15 19 52](https://user-images.githubusercontent.com/2716914/169714235-40697b63-5ddb-4c74-a28c-a851f9851555.png) ![CleanShot 2022-05-22 at 15 20 13](https://user-images.githubusercontent.com/2716914/169714246-a94bcc22-97bc-4a8f-a0a5-9c3225e10ab6.png)
> 
> https://nextcloud.zoeyrae.dev/s/zGFKGkrS8ExxaCy Is the wine.log from the attempt.

Still processing but the flicker in game is fixed mostly, its just even more unplayable. https://www.youtube.com/watch?v=XiKaszUBazY

**Comment by billhollings** [comment]
Here's what I'm seeing in latest MoltenVK under XIV for Mac 3.6.3 on an MacPro M1. What environmental situation can be making it appear to be fixed in my environment, but radically broken on others?

Or put another way...what can I do to cause this to similarly break in my environment? It was definitely similarly broken before the fixes in the current PR.

![Screen Shot 2022-05-22 at 5 15 06 PM](https://user-images.githubusercontent.com/356402/169716654-0d9e36e3-e373-435b-8268-7f90ec66b48e.jpg)

![Screen Shot 2022-05-22 at 5 15 22 PM](https://user-images.githubusercontent.com/356402/169716667-c490cddc-221d-45ee-90ac-19645aa9dfdc.jpg)

![Screen Shot 2022-05-22 at 5 18 48 PM](https://user-images.githubusercontent.com/356402/169716692-deef0c2e-f566-4e64-95da-26b96b0bfe02.jpg)

**Comment by seathasky** [comment]
Interesting. Are you sure that the updated MoltenVK is selected and that it's not using something else on XIV on Mac? @billhollings 
I just tested and getting same results as Zoey

**Comment by ZoeyCluff** [comment]
my game is freshly installed (again) as of this morning. I just made sure my DXVK install was the same as 3.6.5 and it actually got even worse 
![image](https://user-images.githubusercontent.com/2716914/169717196-c7040c4e-cf56-4423-81f4-9cbf3e5537fc.png)
![image](https://user-images.githubusercontent.com/2716914/169717201-f6dfbe0b-9701-45cc-a470-ad1479dbafc6.png)

**Comment by billhollings** [comment]
> Interesting. Are you sure that your updated MoltenVK is selected and that it's not using something else on XIV on Mac?

If I modify code and rebuild, I get different results. I'm very sure I'm running the updated code.

**Comment by Gcenx** [comment]
Then maybe the pushed code doesn’t match your local copy somehow?

**Comment by ZoeyCluff** [comment]
just copying it into the MoltenVK folder in the app doesn't enable it.
Like I overwrite the Stable version of MoltenVK then make sure "use modern MoltenVK" is unchecked in settings.

**Comment by billhollings** [comment]
> Then maybe the pushed code doesn’t match your local copy somehow?

Hmmm. I wonder. Let me check.

**Comment by billhollings** [comment]
> > Then maybe the pushed code doesn’t match your local copy somehow?
> 
> Hmmm. I wonder. Let me check.

Pulling this PR, by sStarting with latest MoltenVK from this repo, then...

```
git checkout -b billhollings-fix-regression-in-dxvk master
git pull https://github.com/billhollings/MoltenVK.git fix-regression-in-dxvk
```

gets me the same (successfully fixed) results that I indicate above. 

Modifying a line of code like this, and rebuilding, breaks it again:

```
bool MVKGraphicsPipeline::isValidVertexBufferIndex(MVKShaderStage stage, uint32_t mtlBufferIndex) {
	return true;
//	return mtlBufferIndex < _descriptorBufferCounts.stages[stage] || mtlBufferIndex > getImplicitBufferIndex(stage, 0);
}
```

BTW...the commits in this PR were changed (squashed) between the first and second versions of this PR. Be sure to start with a clean repo.

**Comment by billhollings** [comment]
> just copying it into the MoltenVK folder in the app doesn't enable it.
> Like I overwrite the Stable version of MoltenVK then make sure "use modern MoltenVK" is unchecked in settings.

In my env, I symlink from both `modern/libMoltenVK.dylib` and  `stable/libMoltenVK.dylib` to my local build pulled from the PR as I describe above.

**Comment by Gcenx** [comment]
Welp good news on my end removed witcher3.dxvk-cache and tested again
![Screen Shot 2022-05-22 at 6 15 32 PM](https://user-images.githubusercontent.com/38226388/169718346-54743359-7cf6-4673-972e-d020e3a8e396.png)

This does seem slower than 1.1.8 though.

**Comment by ZoeyCluff** [comment]
Confirmed, following your steps work
![image](https://user-images.githubusercontent.com/2716914/169718421-acb165e6-4661-4dd2-8231-ba8f9bafd423.png)
FPS Is less than half of what the modern MoltenVK gets though. I usually get 100fps+ in the menus and I was struggling to hit 50fps. In game it's the same.

**Comment by billhollings** [comment]
>Welp good news on my end removed witcher3.dxvk-cache and tested again

> Confirmed, following your steps work

Okay. Great. I'll pull the PR into master then. Thanks for checking caches, etc.

> FPS Is less than half of what the modern MoltenVK gets though. I usually get 100fps+ in the menus and I was struggling to hit 50fps. In game it's the same.

Is this new results? I'll be surprised if this is related to this change. But we can investigate on another issue, if you want to create one after further testing.

**Comment by ZoeyCluff** [comment]
I'm about to work on a new issue for this with more data, but the second I activated 1.1.9 my FPS was halved from what I normally get on  1.1.8 (what i use every day).

**Comment by ZoeyCluff** [comment]
@billhollings Scratch that, my macbook was just being pissy.
![CleanShot 2022-05-22 at 20 08 01](https://user-images.githubusercontent.com/2716914/169725313-47bfd52e-0225-4375-aa40-5d566cedf55a.png)

**Comment by Gcenx** [comment]
@ZoeyCluff still might want to run a long test just to verify before ruling it out entirely.

**Comment by ZoeyCluff** [comment]
Have been since I confirmed it working. I'll be playing a few more hours tonight too.

**Comment by Gcenx** [comment]
@FleetAdmiralButter just a reminder to close this

**Comment by FleetAdmiralButter** [comment]
Resolved with the latest build from master, many thanks for looking into this @billhollings.
